src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.Editor.cs

```diff
@@ -0,0 +1,272 @@
+                _renameType = renameType;
+                _state = state;
+                _service = service;
+                this._moveTypeOptions = moveTypeOptions;
+                this._fromDialog = fromDialog;
+                this._cancellationToken = cancellationToken;
+            }
+
+            private SemanticDocument SemanticDocument => _state.SemanticDocument;
+
+            internal async Task<IEnumerable<CodeActionOperation>> GetOperationsAsync()
+            {
+                var solution = SemanticDocument.Document.Project.Solution;
+
+                if (_renameFile)
+                {
+                    return RenameFileToMatchTypeName(solution);
+                }
+                else if (_renameType)
+                {
+                    return await RenameTypeToMatchFileAsync(solution).ConfigureAwait(false);
+                }
+
+                var documentName = _fromDialog
+                    ? _moveTypeOptions.NewFileName
+                    : _state.TargetFileNameCandidate + _state.TargetFileExtension;
+
+                return await MoveTypeToNewFileAsync(documentName).ConfigureAwait(false);
+            }
+
+            private async Task<IEnumerable<CodeActionOperation>> MoveTypeToNewFileAsync(string documentName)
+            {
+                // fork source document, keep required type/namespace hierarchy and add it to a new document
+                var projectToBeUpdated = SemanticDocument.Document.Project;
+                var newDocumentId = DocumentId.CreateNewId(projectToBeUpdated.Id, documentName);
+
+                var solutionWithNewDocument = await AddNewDocumentWithTypeDeclarationAsync(
+                    SemanticDocument, documentName, newDocumentId, _state.TypeNode, _cancellationToken).ConfigureAwait(false);
+
+                // Get the original source document again, from the latest forked solution.
+                var sourceDocument = solutionWithNewDocument.GetDocument(SemanticDocument.Document.Id);
+
+                // update source document to add partial modifiers to type chain
+                // and/or remove type declaration from original source document.
+                var solutionWithBothDocumentsUpdated = await UpdateSourceDocumentAsync(
+                      sourceDocument, _state.TypeNode, _cancellationToken).ConfigureAwait(false);
+
+                return new CodeActionOperation[] { new ApplyChangesOperation(solutionWithBothDocumentsUpdated) };
+            }
+
+            private async Task<IEnumerable<CodeActionOperation>> RenameTypeToMatchFileAsync(Solution solution)
+            {
+                var symbol = _state.SemanticDocument.SemanticModel.GetDeclaredSymbol(_state.TypeNode, _cancellationToken) as INamedTypeSymbol;
+                var newSolution = await Renamer.RenameSymbolAsync(solution, symbol, _state.DocumentName, SemanticDocument.Document.Options, _cancellationToken).ConfigureAwait(false);
```

Comment:
Should we start an inline rename session instead?  that way if there are errors the user will be notified about them?